src/main/java/build/buildfarm/worker/resources/LocalResourceSetUtils.java

```diff
@@ -37,50 +43,95 @@
 public class LocalResourceSetUtils {
   private static final LocalResourceSetMetrics metrics = new LocalResourceSetMetrics();
 
+  private record SemaphoreLease(String name, int amount, Claim.Stage stage) {}
+
+  private record PoolLease(String name, List<Object> claims, Claim.Stage stage) {}
+
   public static final class LocalResourceSetClaim implements Claim {
-    private final List<Entry<String, Integer>> semaphores;
-    @Getter private final List<Entry<String, List<Integer>>> pools;
+    private final List<SemaphoreLease> semaphores;
+    private final List<PoolLease> pools;
     private final LocalResourceSet resourceSet;
 
     LocalResourceSetClaim(
-        List<Entry<String, Integer>> semaphores,
-        List<Entry<String, List<Integer>>> pools,
-        LocalResourceSet resourceSet) {
+        List<SemaphoreLease> semaphores, List<PoolLease> pools, LocalResourceSet resourceSet) {
       this.semaphores = semaphores;
       this.pools = pools;
       this.resourceSet = resourceSet;
     }
 
+    @Override
+    public synchronized void release(Stage stage) {
+      // only remove resources that match the specified stage
+      // O(n) search for staged resources
+      Iterator<SemaphoreLease> semaphoreIter = semaphores.iterator();
+      while (semaphoreIter.hasNext()) {
+        SemaphoreLease semaphore = semaphoreIter.next();
+        if (semaphore.stage() != stage) {
+          continue;
+        }
+        semaphoreIter.remove();
+        String resourceName = semaphore.name();
+        SemaphoreResource resource = resourceSet.resources.get(resourceName);
+        semaphoreRelease(resource.semaphore(), resourceName, semaphore.amount());
+      }
+      Iterator<PoolLease> poolIter = pools.iterator();
+      while (poolIter.hasNext()) {
+        PoolLease pool = poolIter.next();
+        if (pool.stage() != stage) {
+          continue;
+        }
+        poolIter.remove();
```

Comment:
issue: safe to mutate while iterating?

Reply:
Only safe to mutate with iterator while iterating:
https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/util/Iterator.html
> Iterators allow the caller to remove elements from the underlying collection during the iteration with well-defined semantics.